PERF: quickly resolve inequalities for Quantities with different physical types

neutrinoceros opened the pull request:
### Description
This patch is a candidate optimization (or rebalancing) for `Quantity` comparison (equality/inequality operations),
as was suggested in passing by @mhvk in https://github.com/astropy/astropy/pull/16088#pullrequestreview-1895919657

I used the following script, which heavily draws from the existing benchmark, to easily measure progress
```python
# adapted from astropy-benchmarks (time_quantity_equal)
from time import monotonic_ns
import astropy.units as u
import numpy as np

NLOOPS = 500_000
data = np.array([1.0, 2.0, 3.0])
data = data * u.g

# A different but dimensionally compatible unit
data2 = 0.001 * data * u.kg

data3 = data.copy()

def timeit(code:str, nloops:int=NLOOPS):
    tstart = monotonic_ns()
    for _ in range(nloops):
        exec(code)
    tstop = monotonic_ns()

    elapsed_ms = (tstop-tstart)/1e6
    print(f"'{code}' took {elapsed_ms / 1000 :.1f} s (or {elapsed_ms * 1000 / NLOOPS:.1f} us/iteration)")

timeit("data==data2")
timeit("data!=data2")
timeit("data==data3")
timeit("data!=data3")
```

On `main` I see
```
'data==data2' took 12.7 s (or 25.4 us/iteration)
'data!=data2' took 12.6 s (or 25.1 us/iteration)
'data==data3' took 3.5 s (or 7.0 us/iteration)
'data!=data3' took 3.5 s (or 7.0 us/iteration)
```

An with this patch
```
'data==data2' took 5.5 s (or 11.0 us/iteration)
'data!=data2' took 5.4 s (or 10.9 us/iteration)
'data==data3' took 5.1 s (or 10.2 us/iteration)
'data!=data3' took 5.1 s (or 10.2 us/iteration)
```

On `main`, performance seems biased towards comparison of equal objects (e.g. `data3` and `data`), while on my branch, the time it takes to compare objects is roughly independent of the result, and running the test script is overall faster, *but* the happy path is made longer by about 50%, so there's a rather clear tradeoff and a call to be made: it's not a pure optimization, but more of a re-balancing.

I don't know that the existing imbalance is intended, but maybe it is ?
I also don't know what's the hot path in end-user code (but I'd assume it really depends on applications).
Overall I'm not saying we *should* bring all comparison into the same ballpark performance; my original goal was to
make comparison always faster, or at least never slower, but seeing these interesting results, I figured I might as well
just open the conversation.

I should note that:
- running the test suite locally, I see that it breaks 3 tests, all for the same reason: redshifts' physical type doesn't compare equal to the dimensionless' physical type. After reading through #11202 and glossing over #11204, I'm honestly not sure whether this is a bug or a feature. @nstarman, what do you think ?
- there seem to be only one benchmark for quantity comparison, and it's the one I'm making faster here, so I don't expect existing benchmark to reveal the mixed results I've tried to highlight here, so I'll probably open a PR to add the one we're missing there.


<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented:
@neutrinoceros - note that in your example (at least as written) the unit of `data2` is "g kg", so the unit is not dimensionally compatible (I expect this is an error with adding the unit to `data` later?). I think we should definitely prioritize comparisons with compatible units...

mhvk commented:
p.s. Bit more generally, I think it would be best to focus on speeding up (if possible), `_to_own_unit()` (and hence `to_value()`) since those are used very commonly elsewhere as well.

nstarman commented:
> running the test suite locally, I see that it breaks 3 tests, all for the same reason: redshifts' physical type doesn't compare equal to the dimensionless' physical type

Yes, this is a deliberate choice. Redshift has a physical meaning related to the distance. The primary reason for having a redshift unit is to track factors of redshift across a calculation and distinguish it from normal dimensionless numbers.

nstarman commented on astropy/units/quantity.py:
```diff
@@ -1138,6 +1138,13 @@ def get_virtual_unit_attribute():
     def __eq__(self, other):
+        if (
+            hasattr(other, "unit")
+            and hasattr(other.unit, "physical_type")
+            and (other.unit.physical_type) != self.unit.physical_type
```
Is this faster?

```suggestion
            and getattr(other.unit, "physical_type", True) != getattr(self.unit, "physical_type", False)
```

neutrinoceros replied:
It *is* faster. Running my script again:
my patch:
```
'data==data2' took 5.7 s (or 11.4 us/iteration)
'data!=data2' took 5.6 s (or 11.3 us/iteration)
'data==data3' took 5.2 s (or 10.4 us/iteration)
'data!=data3' took 5.4 s (or 10.8 us/iteration)
```
with your suggestion
```
'data==data2' took 4.6 s (or 9.1 us/iteration)
'data!=data2' took 4.5 s (or 9.0 us/iteration)
'data==data3' took 4.7 s (or 9.3 us/iteration)
'data!=data3' took 4.6 s (or 9.3 us/iteration)
```
I'll push this, thank you !

neutrinoceros commented:
> note that in your example (at least as written) the unit of data2 is "g kg", so the unit is not dimensionally compatible (I expect this is an error with adding the unit to data later?).
> I think we should definitely prioritize comparisons with compatible units...

This may be a mistake, but I took it as is from the benchmark code:

https://github.com/astropy/astropy-benchmarks/blob/c54d814aef79930d5cf6036d89c7c16c16d7740e/benchmarks/units.py#L159-L164

I think this case is actually worth a benchmark, but we clearly need to have another one in the mix too. I'll open a PR today.

> I think it would be best to focus on speeding up (if possible), _to_own_unit() (and hence to_value()) since those are used very commonly elsewhere as well.

Perhaps I should have told the broader story in the PR description: this actually was my original goal, until I discovered that a lot of the logic to distinguish between compatible and incompatible units was built using `try/except` blocks, which are the root cause of the performance imbalance I'm revealing. I expect that refactoring this logic to not rely on exception catching would be very difficult, so I tried to work on a different level instead, hence this patch.

>  The primary reason for having a redshift unit is to track factors of redshift across a calculation and distinguish it from normal dimensionless numbers.

I'm slightly confused. Then why was it decided to allow redshift quantites to compare equal to dimensionless ones ? Isn't there a conflict between these two choices ?

neutrinoceros commented:
Also, while away from keyboard, I wanted to know the tipping point between the reference implementation and my branch, in terms of the relative abundance of hits VS misses (where I define a hit as a comparison between equal objects and a miss as a comparison between unequal objects). The answer I found is about 4.5. In other words, when 18% (or more) of all comparisons being made are misses, my branch is overall faster. This seems low enough that I can phantom most applications would benefit from it.

neutrinoceros commented:
And refactoring futher to skip `PhysicalType.__eq__` we get:
```
'data==data2' took 3.1 s (or 6.2 us/iteration)
'data!=data2' took 3.1 s (or 6.2 us/iteration)
'data==data3' took 3.8 s (or 7.6 us/iteration)
'data!=data3' took 3.8 s (or 7.6 us/iteration)
```

further moving the tipping point to ~32, meaning just 3% misses is sufficient to make the patch an overall gain.
With this version, the happy path is only degraded by about 9%, while the "unhappy" path is improved by more than 4x
Looks to me this is getting less controversial by the minute 😸

mhvk commented:
@neutrinoceros - I think you overestimate the importance of unequal units in `Quantity.__eq__` - I would think that in almost all cases, that is a problem with user input and hence soon enough an exception will be raised.  I suspect a large fraction is comparisons with equivalent units where the resulting array of bool is used to index or so (though, given floats, that is perhaps more likely for greater than, etc., except for `== 0`. (Note that those code paths will do badly regardless with `__eq__` returning a plain `False` - we never properly followed numpy's lead in ensuring that `__eq__` always returns an array with broadcasted shape...).

One possible way to test this is to count results in the astropy test suite (though there again I'd be wary of tests that explicitly check wrong user input...)

Overall, I'm very wary of performance improvements that work by adding code. For instance, in this case, if python manages to make the except path faster, any `if`s beforehand will just make the code slower (and this is not so unrealistic: it is only a few versions ago that python ensured that the success path of a `try/except` has no cost).

More specific to this PR - it seems the main gain is from making unit comparisons faster - that seems very worthwhile on its own, but might have wider benefit if done in the units code itself. Though it is hard to be sure without doing a bit more detailed profiling.

neutrinoceros commented:
I just realized that my approach was intrinsically biased towards unequal cases because of a mistake I made while reading/adapting the benchmark code 🤦🏻‍♂️ 
In actuality, the benchmark is really comparing quantities that are dimensionally compatible, I just broke this property as I extracted the code from a benchmark class and dropped the distinction between two variables (`self.data` and `data`).

So, thanks a lot for being so wary, it's clearly pushed me in the right direction of critical thinking 😅
I also agree that it'd be best to optimize `Unit.__eq__` instead of `Quantity.__eq__` if possible, since the former implies the latter. Anyway, now I see that this specific patch is probably not going anywhere, so I'll close it and start over. Thank the both of you for replying so promptly !
